### PR TITLE
[SYCL][Graph] Add implicit queue recording mechanism

### PR DESCRIPTION
**Pull request by fabiomestre**
Updates the semantics of the record & replay API to allow queues to be
implicitly set to recording mode when submitting command-groups
that contains dependencies from other recording queues.

### CODE REVIEW

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 1]
```diff
@@ -1453,7 +1453,18 @@ graph LR
...
+
+Submitting a command-group to a queue can implicitly change its state
+to `queue_state::recording`. This will occur when the command-group depends on
+an event that has been returned by a queue in the recording state.
+
+A queue whose state has been set to `queue_state::recording` using this
+mechanism, will behave as if it had been passed as an argument to
+`command_graph::begin_recording()`. In particular, its state will not
+change again until `command_graph::end_recording()` is called.
+
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:1468]: I think adding a code snippet to illustrate this behaviour would be useful to the reader, something like: 
```cpp
// User begins recording on q1
graph.begin_recording(q1);
// Node added to graph by recording queue submission
auto e1 = q1.parallel_for(...);
  
// Dependency on `e1` from queue `q1` in recording state causes `q2` to enter recording mode and this
// `e2` command to be added as a node in the graph with an edge on `e1`
auto e2 = q2.parallel_for(e1, ...);

// ends recording on q1 and q2
graph.end_recording();
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 2]
```diff
@@ -1453,7 +1453,18 @@ graph LR
     Recording -->|End Recording| Executing
 ....
 
-==== Queue Properties
+==== Implicit Queue Recording
+
+Submitting a command-group to a queue can implicitly change its state
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: to a queue in the **executable state** can implicitly ...

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 3]
```diff
@@ -1453,7 +1453,18 @@ graph LR
...
+to `queue_state::recording`. This will occur when the command-group depends on
+an event that has been returned by a queue in the recording state.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Think we should add an extra sentence clarifying that this takes effect starting on that command-group submission not after it, so the command-group is added as a node in the graph with an edge to that node the dependent event came from.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 4]
```diff
@@ -1453,7 +1453,18 @@ graph LR
     Recording -->|End Recording| Executing
 ....
 
-==== Queue Properties
+==== Implicit Queue Recording
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Nitpick, I'm not sure if "Transitive" or "Implicit" is a better term to describe this behavior

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 5]
```diff
@@ -1453,7 +1453,39 @@ graph LR
...
+// Since there is a dependency on e1 which was created by a queue being
+// recorded, q2 immediately enters record mode, and a new node is created
+// with an edge between e1 and e2.
+auto e2 = q2.single_task(e1, ...);
+
+// Ends recording on q1 and q2.
+graph.end_recording();
+----
+
+==== Queue Properties;
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
==== Queue Properties
```

**Review by EwanC — APPROVED** [review]

**Comment by fabiomestre** [comment]
Upstream PR: https://github.com/intel/llvm/pull/14453

**Pull request closed without merging**